File: scripts/memscan.c
/*
 * memscan — fast float memory scanner for /proc/PID/mem
 * Usage: memscan <pid> <float_value> [tolerance] [--heap]
 * --heap: only scan scudo heap + cocos2d regions (5x faster)
 * Output: hex_address float_value (one per line)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <fcntl.h>
#include <unistd.h>

#define BUF_SIZE (4 * 1024 * 1024)  /* 4 MB read buffer */
#define MAX_MATCHES 100000

struct match {
    unsigned long addr;
    float val;
};

static struct match matches[MAX_MATCHES];
static int n_matches = 0;
static char buf[BUF_SIZE];

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: memscan <pid> <value> [tolerance] [--heap]\n");
        return 1;
    }

    int pid = atoi(argv[1]);
    float target = atof(argv[2]);
    float tol = argc > 3 && argv[3][0] != '-' ? atof(argv[3]) : 1.0f;

    int heap_only = 0;
    for (int i = 3; i < argc; i++) {
        if (strcmp(argv[i], "--heap") == 0) heap_only = 1;
    }

    /* Open maps */
    char path[256];
    snprintf(path, sizeof(path), "/proc/%d/maps", pid);
    FILE *maps = fopen(path, "r");
    if (!maps) { perror("fopen maps"); return 1; }

    /* Open mem */
    snprintf(path, sizeof(path), "/proc/%d/mem", pid);
    int fd = open(path, O_RDONLY);
    if (fd < 0) { perror("open mem"); return 1; }

    char line[512];
    long total_scanned = 0;
    int regions = 0;

    while (fgets(line, sizeof(line), maps)) {
        unsigned long start, end;
        char perms[8];
        if (sscanf(line, "%lx-%lx %4s", &start, &end, perms) != 3)
            continue;

        /* Only writable regions */
        if (perms[1] != 'w') continue;

        long size = end - start;
        /* Skip tiny and huge regions */
        if (size < 4096 || size > 50000000) continue;

        /* Region name filtering */
        char *name = strrchr(line, ' ');
        if (name) name++; else name = "";

        /* Always skip device and system mappings */
        if (strstr(name, "/dev/") || strstr(name, "/system/"))
            continue;

        if (heap_only) {
            /* Only scan: scudo heap, cocos2d lib, fingersoft lib, [heap] */
            int dominated = 0;
            if (strstr(name, "scudo")) dominated = 1;
            if (strstr(name, "cocos2d")) dominated = 1;
            if (strstr(name, "fingersoft")) dominated = 1;
            if (strstr(name, "[heap]")) dominated = 1;
            if (strstr(name, "[anon]") || name[0] == '\n' || name[0] == '\0')
                dominated = 1;  /* unnamed anonymous regions */
            if (!dominated) continue;
        }

        regions++;

        /* Read in chunks */
        long offset = 0;
        while (offset < size && n_matches < MAX_MATCHES) {
            long to_read = size - offset;
            if (to_read > BUF_SIZE) to_read = BUF_SIZE;

            ssize_t nread = pread(fd, buf, to_read, start + offset);
            if (nread < 4) break;

            total_scanned += nread;

            /* Scan for float matches */
            for (long i = 0; i <= nread - 4; i += 4) {
                float val;
                memcpy(&val, buf + i, 4);
                /* Skip NaN/Inf */
                if (val != val || val == 1.0f/0.0f || val == -1.0f/0.0f)
                    continue;
                if (fabsf(val - target) < tol) {
                    if (n_matches < MAX_MATCHES) {
                        matches[n_matches].addr = start + offset + i;
                        matches[n_matches].val = val;
                        n_matches++;
                    }
                }
            }
            offset += nread;
        }
    }

    fclose(maps);
    close(fd);

    fprintf(stderr, "Scanned %ld MB in %d regions, found %d matches\n",
            total_scanned / (1024*1024), regions, n_matches);

    for (int i = 0; i < n_matches; i++) {
        printf("0x%016lx %.4f\n", matches[i].addr, matches[i].val);
    }

    return 0;
}

File: scripts/memscan.py
#!/usr/bin/env python3
"""Memory scanner for HCR2 — find float values in game process memory."""

import struct
import subprocess
import sys
import time


def get_pid(container: str) -> int:
    r = subprocess.run(
        ["docker", "exec", container, "pidof", "com.fingersoft.hcr2"],
        capture_output=True, timeout=5,
    )
    return int(r.stdout.decode().strip())


def get_heap_regions(container: str, pid: int) -> list[tuple[int, int]]:
    """Get writable heap regions (scudo allocator + anon)."""
    r = subprocess.run(
        ["docker", "exec", container, "cat", f"/proc/{pid}/maps"],
        capture_output=True, timeout=5,
    )
    regions = []
    for line in r.stdout.decode().splitlines():
        parts = line.split()
        perms = parts[1]
        if "w" not in perms:
            continue
        start, end = (int(x, 16) for x in parts[0].split("-"))
        size = end - start
        if size > 50_000_000 or size < 4096:
            continue
        name = parts[-1] if len(parts) > 5 else ""
        if "/dev/" in name or "/system/" in name:
            continue
        regions.append((start, end))
    return regions


def read_mem(container: str, pid: int, start: int, size: int) -> bytes:
    """Read memory chunk via dd on /proc/PID/mem."""
    cmd = (
        f"dd if=/proc/{pid}/mem bs=1 skip={start} count={size} 2>/dev/null"
    )
    r = subprocess.run(
        ["docker", "exec", container, "sh", "-c", cmd],
        capture_output=True, timeout=10,
    )
    return r.stdout


def scan_float(container: str, pid: int, target: float, tolerance: float = 0.5) -> list[int]:
    """Scan all heap for float value within tolerance."""
    regions = get_heap_regions(container, pid)
    target_bytes = struct.pack("<f", target)
    matches = []
    total_scanned = 0

    for start, end in regions:
        size = end - start
        if size > 10_000_000:  # skip very large for speed
            continue
        try:
            data = read_mem(container, pid, start, size)
        except Exception:
            continue
        if len(data) < 4:
            continue
        total_scanned += len(data)

        # Scan for float matches
        for offset in range(0, len(data) - 3, 4):
            val = struct.unpack_from("<f", data, offset)[0]
            if abs(val - target) < tolerance:
                addr = start + offset
                matches.append((addr, val))

    print(f"Scanned {total_scanned / 1024 / 1024:.1f} MB, found {len(matches)} matches")
    return matches


def rescan(container: str, pid: int, addrs: list[int], target: float, tolerance: float = 0.5) -> list[int]:
    """Re-read specific addresses and filter by new target value."""
    remaining = []
    for addr, _ in addrs:
        try:
            data = read_mem(container, pid, addr, 4)
            if len(data) == 4:
                val = struct.unpack("<f", data)[0]
                if abs(val - target) < tolerance:
                    remaining.append((addr, val))
        except Exception:
            continue
    print(f"Rescan: {len(remaining)}/{len(addrs)} remain for target={target}")
    return remaining


if __name__ == "__main__":
    container = sys.argv[1] if len(sys.argv) > 1 else "hcr2-0"
    target = float(sys.argv[2]) if len(sys.argv) > 2 else 181.0

    pid = get_pid(container)
    print(f"PID: {pid}, scanning for float {target}...")

    t0 = time.time()
    matches = scan_float(container, pid, target, tolerance=1.0)
    print(f"Scan took {time.time() - t0:.1f}s")

    if matches:
        print(f"\nFirst 20 matches:")
        for addr, val in matches[:20]:
            print(f"  0x{addr:016x} = {val:.2f}")
